File: 39-combination-sum/39-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = set()
        perm = []
        local = defaultdict(int)
        candidateFreqs = defaultdict(set)
        def backtrack():
            if sum(perm) == target:
                res.add(tuple(perm[:]))
                return
            elif sum(perm) > target:
                return
            for i in range(len(candidates)):
                if len(perm) == 0 or candidates[i] >= perm[-1]:
                    perm.append(candidates[i])
                    local[candidates[i]] += 1
                    backtrack()
                    perm.pop()
                    local[candidates[i]] -= 1
        backtrack()
        return list(res)

File: 39-combination-sum/39-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res = {};
        vector<int> subset = {};
        combinationSumHelper(candidates, 0, target, subset, res);
        return res;
    }
    
    void combinationSumHelper(vector<int>& candidates, int start, int curr, vector<int> &subset, vector<vector<int>>& result){
        if(curr == 0){
            result.push_back(subset);
        }else if(curr < 0){
            return;
        }
        
        for(int i = start; i < candidates.size(); ++i){
            subset.push_back(candidates[i]);
            combinationSumHelper(candidates, i, curr - candidates[i], subset, result);
            subset.pop_back();
        }
    }
};